pkg/skaffold/initializer/init_test.go

```diff
@@ -59,3 +61,120 @@ func TestPrintAnalyzeJSON(t *testing.T) {
 		})
 	}
 }
+
+func TestWalk(t *testing.T) {
+	emptyFile := ""
+	tests := []struct {
+		name                string
+		filesWithContents   map[string]string
+		expectedConfigs     []string
+		expectedDockerfiles []string
+		force               bool
+		shouldErr           bool
+	}{
+		{
+			name: "should return correct k8 configs and dockerfiles",
+			filesWithContents: map[string]string{
+				"config/test.yaml":  emptyFile,
+				"k8pod.yml":         emptyFile,
+				"README":            emptyFile,
+				"deploy/Dockerfile": emptyFile,
+				"Dockerfile":        emptyFile,
+			},
+			force: false,
+			expectedConfigs: []string{
+				"config/test.yaml",
+				"k8pod.yml",
+			},
+			expectedDockerfiles: []string{
+				"Dockerfile",
+				"deploy/Dockerfile",
+			},
+			shouldErr: false,
+		},
+		{
+			name: "should skip hidden dir",
+			filesWithContents: map[string]string{
+				".hidden/test.yaml":  emptyFile,
+				"k8pod.yml":          emptyFile,
+				"README":             emptyFile,
+				".hidden/Dockerfile": emptyFile,
+				"Dockerfile":         emptyFile,
+			},
+			force: false,
+			expectedConfigs: []string{
+				"k8pod.yml",
+			},
+			expectedDockerfiles: []string{
+				"Dockerfile",
+			},
+			shouldErr: false,
+		},
+		{
+			name: "should not error when skaffold.config present and force = true",
+			filesWithContents: map[string]string{
+				"skaffold.yaml": `apiVersion: skaffold/v1beta6
+kind: Config
+deploy:
+  kustomize: {}`,
+				"config/test.yaml":  emptyFile,
+				"k8pod.yml":         emptyFile,
+				"README":            emptyFile,
+				"deploy/Dockerfile": emptyFile,
+				"Dockerfile":        emptyFile,
+			},
+			force: true,
+			expectedConfigs: []string{
+				"config/test.yaml",
+				"k8pod.yml",
+			},
+			expectedDockerfiles: []string{
+				"Dockerfile",
+				"deploy/Dockerfile",
+			},
+			shouldErr: false,
+		},
+		{
+			name: "should  error when skaffold.config present and force = false",
+			filesWithContents: map[string]string{
+				"config/test.yaml":  emptyFile,
+				"k8pod.yml":         emptyFile,
+				"README":            emptyFile,
+				"deploy/Dockerfile": emptyFile,
+				"Dockerfile":        emptyFile,
+				"skaffold.yaml": `apiVersion: skaffold/v1beta6
+kind: Config
+deploy:
+  kustomize: {}`,
+			},
+			force:               false,
+			expectedConfigs:     nil,
+			expectedDockerfiles: nil,
+			shouldErr:           true,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			testDir, deleteFunc := testutil.NewTempDir(t)
+			rootDir := testDir.Root()
+			deleteFunc()
+			writeAllFiles(testDir, test.filesWithContents)
+			potentialConfigs, dockerfiles, err := walk(rootDir, test.force, testValidDocker)
+			testutil.CheckErrorAndDeepEqual(t, test.shouldErr, err,
+				testDir.Paths(test.expectedConfigs), potentialConfigs)
+			testutil.CheckErrorAndDeepEqual(t, test.shouldErr, err,
+				testDir.Paths(test.expectedDockerfiles), dockerfiles)
+			os.Remove(rootDir)
```

Comment:
This is not needed.

Reply:
One point that I never mentioned to other members of the team but that I should have:
I find it much easier to read when there's some kind of structure in a test:
```
setup1()
setup2()

action()

assertions()
```

When we add a space before and after the action, it creates a nicer structure.

Reply:
If we create a style guide, we should put all those preferences in.

Reply:
umm actually i have this point of view. I don't remember which testing doc it was, but there were few points which stood out for me
1.  Repetition of code in tests is ok instead of re-direction in another file or method. Having long test function so that all code is contained in the test case instead of having them someother method like `action()` and `assertions()` especially where `action()` and `assertions` is the crux of the code. (Code Proximity Vs Modular approach)
```
setup1()
setup()

// do the action here

// test the assertions here

```
2. Make test case input clear.
To avoid guesswork,  write all the input values.
```
should = false  // though by default bool are false.
```
3. Only refactor code and convert it into a helper function, if its used in more than place.

WDYT?

Reply:
Actually, it could be i read this on TOT :) maybe i can find those and we can discuss them.